_posts/2021-02-05-solidity-fuzzer-update.md

```diff
@@ -0,0 +1,161 @@
+
+If you are curious as to how the fuzzer may have "synthesised" the NatSpec comment, one of the ways (caution: pure speculation) is the following:
+  - In some fuzzer run `N`, adding dictionary token "//" (i.e., comment syntax in Solidity)
+  - In run `N' > N`, adding dictionary token "return"
+  - Finally, arriving at `///@return` through a sequence of additions/shuffling.
+
+The Solidity fuzzing dictionary [may be found here][5].
+If you spot omissions (especially those that may yield interesting mutants), please let us know.
+PRs welcome :-)
+
+Notes:
+
+This [issue][4] is a few months old has been fixed in `0.7.6`.
+The issue is that Solidity modifiers don't have return variables, and hence a `@return` NatSpec tag does not make sense.
+The program above should have resulted in a parsing error (which is what the fix ensures) but instead resulted in a failed assertion.
+
+#### Back-end Fuzzers
+
+In order to test if code generation is sound, valid program (input) needs to be fed to the compiler.
+If the program contains syntactic (`iff` instead of `if`) or semantic (use undeclared identifier) errors, it won't be accepted by the compiler.
+Semantic fuzzers are designed to ensure program input does not contain such errors.
+
+To implement semantic fuzzers, we rely on Google's [libprotobuf-mutator][2] library.
+Broadly speaking, writing a semantic fuzzer involves the following steps:
+  - Write a (context-free) grammar for Solidity in the protobuf interface definition language.
+  - Implement a converter interface that accepts a generation of the protobuf grammar and converts into target language
+
+We have implemented one such semantic fuzzer for testing the correctness of the Yul optimiser.
+The fuzzer randomly generates Yul programs whose behavior is tracked pre and post optimisation in order to flag stateful divergence.
+This helps us determine if a bug was introduced into the optimiser.
+
+##### Bug found by back-end fuzzer
+
+The redundant assignment eliminator is a Yul optimiser step that (surpise!) removes redundant assignments to a variable.
+The back-end fuzzer generated the following example that demonstrated an error in this step:
+
+```
+{
+        for {let i:= 0} lt(i,2) {i := add(i,1)}
+        {
+                // x is declared and implicitly
+                // initialized to zero.
+                let x
+                // This assignment is not redundant
+                // since it is used by mstore statement
+                // that follows the if statement
+                x := 1337
+                if lt(i,1) {
+                        // This assignment is redundant
+                        x := 42
+                        continue
+                }
+                mstore(0, x)
+        }
+}
+```
+
+The redundant assignment eliminator eliminated all assignments to `x` leading to `mstore(0, x)` thereby optimising the program to one containing `mstore(0, 0)` instead of `mstore(0, 1337)`.
```

Comment:
```suggestion
The redundant assignment eliminator eliminated all assignments to `x` leading to `mstore(0, x)`, thereby optimising the program to one containing `mstore(0, 0)` instead of `mstore(0, 1337)`.
```